Applications/Spire/Include/Spire/Ui/RecentColors.hpp

```diff
@@ -0,0 +1,44 @@
+#ifndef SPIRE_RECENT_COLORS_HPP
+#define SPIRE_RECENT_COLORS_HPP
+#include <array>
+#include <deque>
+#include <QColor>
+
+namespace Spire {
+
+  //! Represents a data structure for storing recently selected colors in a
+  //  FIFO manner.
+  class RecentColors {
+    public:
+
+      //! The maximum number of recent colors to store.
+      static const int RECENT_COLOR_COUNT = 9;
+
+      //! Returns the default recent colors.
+      static const RecentColors& get_default_colors();
+
+      RecentColors() = default;
+
+      //! Inserts a color into the recently selected colors.
+      /*!
+        \param color The color to insert.
+      */
+      void add_color(const QColor& color);
+
+      //! Returns the collection of recently selected colors.
+      std::array<QColor, RECENT_COLOR_COUNT> get_colors() const;
+
+      //! Returns true if two RecentColors share the same recently selected
+      //! colors in the same order.
+      bool operator ==(const RecentColors& rhs) const;
+
+      //! Returns false if two RecentColors do not share the same recently
+      //! selected colors in the same order.
+      bool operator !=(const RecentColors& rhs) const;
+
+    private:
+      std::deque<QColor> m_recent_colors;
```

Comment:
Use a combination of a hash set (probably `QSet`) and `std::list` instead. Your `add_color` should then be constant time.
It is more of an exercise than an actual performance consideration with the container size of 9, but it communicates the intention more clearly once you are used to such algorithms.

Reply:
Wouldn't a set lose the ordering?

Reply:
It certainly will. The idea is that in your algorithm, you perform the following four operations:

1. Check for existence of an element and locate it if it does exist.
2. Remove the element located if any.
3. Pop the first element.
4. Push a new element to the end of the container.

`std::deque` allows you to achieve O(1) for (3) and (4), but both (1) and (2) are linear as it behaves largely as a normal array.
Now, if you simply replace `std::deque` with `std::list`, you are able to achieve O(1) for (2), while keeping the constant complexity of (3) and (4). The complexity of (1) remains linear though, as you have to traverse the entire list to look for the element. As a single `add_color` operations needs to at least perform (1), one would argue that the overall complexity is still linear.

Finally, you are able to add a supplementary structure to make (1) constant time as well. The thinking is: for fast existence/absence tests, as well for the fast location elements, we use hash tables. If you maintain a table from element keys (`QColor`) to list iterators, you are then able to locate an existing element in constant time, and then remove it from the list (in constant time as well). In other words, the ordering is maintained by the list, while the hash serves as an index for that list.

To correct myself, you probably need `QHash` (or `std::unordered_map`, if `std::hash` is implemented for `QColor`) instead of `QSet` to store list iterators as satellite data.

As mentioned before, you are unlikely to benefit from constant complexity in this case, yet this is a repetitive pattern and I have come across similar sub-optimal implementation. The main goal is to understand how you can benefit from using other data structures in situations like this, and to avoid poor implementations in less trivial scenarios.

Reply:
Keep the implementation as is. Increasing the complexity by adding a hash and a list in order to keep track of 9 colors is simply overkill.

Reply:
If any change is to be made it would be to simplify it even further and just use an ```std::array```.

Reply:
I don't understand how point (2) is solved with the iterator hashes; wouldn't it have to be 'refreshed' every time the list is modified?

Reply:
(2) is solved with the use of a linked list rather than the hash. It pays to understand why pointer removals from linked lists are constant.
List modifications do not invalidate existing iterators (except for removed elements). If you have an iterator to, say, 4th element in a list and then remove the 2nd element, the iterator to the 4th element is unchanged. Similarly to the case of node removal, it is a good idea to understand why the iterators stay the same, as well as how arrays and linked lists are different at all.